Sources/Rendering/WebGPU/UniformBuffer/index.js

```diff
@@ -0,0 +1,338 @@
+    }
+  };
+
+  publicAPI.setValue = (name, val) => {
+    publicAPI.sortBufferEntries();
+    const idx = model.bufferEntryNames.get(name);
+    if (idx === undefined) {
+      vtkErrorMacro(`entry named ${name} not found in UBO`);
+      return;
+    }
+    const entry = model.bufferEntries[idx];
+    publicAPI.createView(entry.nativeType);
+    const view = model[entry.nativeType];
+    if (view[entry.offset / view.BYTES_PER_ELEMENT] !== val) {
+      view[entry.offset / view.BYTES_PER_ELEMENT] = val;
+      model.sendDirty = true;
+    }
+  };
+
+  publicAPI.setArray = (name, arr) => {
+    publicAPI.sortBufferEntries();
+    const idx = model.bufferEntryNames.get(name);
+    if (idx === undefined) {
+      vtkErrorMacro(`entry named ${name} not found in UBO`);
+      return;
+    }
+    const entry = model.bufferEntries[idx];
+    publicAPI.createView(entry.nativeType);
+    const view = model[entry.nativeType];
+    for (let i = 0; i < arr.length; i++) {
+      if (view[entry.offset / view.BYTES_PER_ELEMENT + i] !== arr[i]) {
+        view[entry.offset / view.BYTES_PER_ELEMENT + i] = arr[i];
+        model.sendDirty = true;
+      }
+    }
+  };
+
+  publicAPI.getNativeView = (name) => {
+    publicAPI.sortBufferEntries();
+    const idx = model.bufferEntryNames.get(name);
+    if (idx === undefined) {
+      vtkErrorMacro(`entry named ${name} not found in UBO`);
+      return undefined;
+    }
+    const entry = model.bufferEntries[idx];
+    publicAPI.createView(entry.nativeType);
+    const bpe = model[entry.nativeType].BYTES_PER_ELEMENT;
+    model.sendDirty = true;
+    return new window[entry.nativeType](
+      model.arrayBuffer,
+      entry.offset,
+      entry.sizeInBytes / bpe
+    );
+  };
+
+  publicAPI.getSendTime = () => model.sendTime.getMTime();
+  publicAPI.getShaderCode = () => {
+    // sort the entries
+    publicAPI.sortBufferEntries();
+
+    let result = `[[block]] struct ${model.name}Struct\n{\n`;
+    for (let i = 0; i < model.bufferEntries.length; i++) {
+      const entry = model.bufferEntries[i];
+      result += `[[offset(${entry.offset})]] ${entry.name}: ${entry.type};\n`;
+    }
+    result += `};\n[[binding(${model.binding}), group(${model.group})]] var<uniform> ${model.name}: ${model.name}Struct;`;
+    return result;
+  };
+}
+
+// ----------------------------------------------------------------------------
+// Object factory
+// ----------------------------------------------------------------------------
+
+const DEFAULT_VALUES = {
+  bufferEntries: null,
+  bufferEntryNames: null,
+  sizeInBytes: 0,
+  name: null,
+  binding: 0,
+  group: 0,
+};
+
+// ----------------------------------------------------------------------------
+
+export function extend(publicAPI, model, initialValues = {}) {
+  Object.assign(model, DEFAULT_VALUES, initialValues);
+
+  // Build VTK API
+  macro.obj(publicAPI, model);
+
+  // Internal objects
+  model.bufferEntryNames = new Map();
```

Comment:
ideally no`Map()` in model. Can this just be a const fully private? (in function rather than extend)?
Otherwise for this type of classes we don't care as we won't ask for `.getState()`...

Reply:
Hmm, Map is a really useful class to me and using an Object as a map seems like bad practice for a few reasons. While I can make Maps private that really kills the ability for people to subclass and do things that require access to that member. So... since this will continue to come up, how would you like me to implement protected Maps such that they are not private?

Reply:
I'm not against using `Map()`, I'm just against binding it to model if not mandatory (accessible for sub-classes).

Reply:
you can create a model._protected = { ... all the protected objects }

Reply:
scratch that... Just put it inside the model with a `_` as prefix. I guess the idea would be that things in the model with an `_` should not have a getter and should not be serialized. That way we can update the `getState` to take into account that aspect.

Reply:
The linter will bother you on that but you can turn it off for that part specifically

Reply:
If we are going to use that as a convention in vtk.js then we should add it to our .eslintrc.js yes?

Reply:
yes let's add that to our eslintrs